llvm/lib/Target/SystemZ/SystemZISelLowering.cpp

```diff
@@ -7223,7 +7223,16 @@ SDValue SystemZTargetLowering::combineSTORE(
       if (C->getAPIntValue().getBitWidth() > 64 || C->isAllOnes() ||
           isInt<16>(C->getSExtValue()) || MemVT.getStoreSize() <= 2)
         return;
-      SystemZVectorConstantInfo VCI(APInt(TotBytes * 8, C->getZExtValue()));
+
+      APInt Val = C->getAPIntValue();
+      // Truncate Val in case of a truncating store.
+      if (!llvm::isUIntN(TotBytes * 8, Val.getZExtValue())) {
+        assert(SN->isTruncatingStore() &&
+               "Non-truncating store and immediate value does not fit?");
+        Val = Val.trunc(TotBytes * 8);
+      }
+
+      SystemZVectorConstantInfo VCI(Val);
```

Comment:
This still uses the size of Val rather than TotBytes, so it still pessimizes some cases (where the original Val.getExtValue() *does* fit, like in my example of a 64->32 truncating store of 0x01010101, which should use VREPI).

Reply:
sorry, was just about to change that.

I guess in the non-truncating case the original APInt should be of the matching bitwidth, but it doesn't hurt to be explicit and use TotBytes * 8. 

And in the truncating case, it would be truncated where also the Bitwidth of the new APInt should be reduced.

Reply:
and yes, if there is a truncating store with a value that fits, this also makes a difference.